### PR TITLE
[chore] github action to publish

### PR DESCRIPTION
**Pull request by erin-at-work**
### *Summary*

- Add publishing workflow

**DO NOT MERGE:** This PR was initiated to start a conversation on our current deployment workflow.

### CODE REVIEW

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      - name: Update version
+        run: yarn version --${{ env.PACKAGE_VERSION }}
+
+      - name: Build npm pkge
+        run: yarn build:prod
+
+      - name: Publish
+        run: npm publish --tag ${{ env.DIST_TAG }}
+        env:
+          NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
```
Comment (erin-at-work) [inline .github/workflows/publish.yml:59]: TODO: add secrets

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+          node-version-file: ".nvmrc"
+          cache: 'yarn'
+
+      - name: Install
+        run: yarn install --frozen-lockfile
+
+      - name: Set registry url
+        uses: actions/setup-node@v2
+        with:
+          registry-url: 'https://registry.npmjs.org'
```
Comment (ljharb) [inline .github/workflows/publish.yml:51]: is this needed? this registry is the default.

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      DIST_TAG: ${{ github.event.inputs.distTag }}
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Use Node.js
+        uses: actions/setup-node@v2
+        with:
+          node-version-file: ".nvmrc"
```
Comment (ljharb) [inline .github/workflows/publish.yml:42]: heads up that setup-node doesn't actually support everything that `.nvmrc` does; i suggest using my ljharb/node/install action instead.

File: .github/workflows/publish.yml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+  
+  release:
+    name: Publish to Github Packages
+    runs-on: ubuntu-latest
+    needs: [authorize]
+    env:
+      PACKAGE_VERSION: ${{ github.event.inputs.packageVersion }}
+      DIST_TAG: ${{ github.event.inputs.distTag }}
```
Comment (ljharb) [inline .github/workflows/publish.yml:33]: why are we publishing to github packages? we should only be publishing to npm.

Comment (erin-at-work) [reply .github/workflows/publish.yml:33]: Typo, should've read NPM packages

**Review by jeongmincho — APPROVED** [review]
plz ignore me - testing my DUO app

**Comment by cb-heimdall** [comment]
Review Error for jeongmincho @ 2022-04-07 17:49:15 UTC
User failed mfa authentication, see [go/mfa-help](http://go/mfa-help)

**Review by jeongmincho — APPROVED** [review]
i think this is it

**Comment by mikehunter1990** [comment]
How

**Comment by Newman1515** [comment]
diff --git a/.github/workflows/publish.yml b/.github/workflows/publish.yml
new file mode 100644
index 00000000..0fc25bd1
--- /dev/null
+++ b/.github/workflows/publish.yml
@@ -0,0 +1,62 @@
+name: Publish Package to npm
+
+on:
+  workflow_dispatch:
+    inputs:
+      packageVersion:
+        description: "The version to publish (e.g. patch, minor, major)"
+        required: true
+        default: "patch"
+      distTag:
+        description: "The dist-tag to publish (e.g. latest, beta)"
+        required: true
+        default: "latest"
+
+jobs:
+  authorize:
+    name: Authorize
+    runs-on: ubuntu-latest
+    steps:
+      - name: ${{ github.actor }} permission check to do a release
+        uses: "lannonbr/repo-permission-check-action@2.0.2"
+        with:
+          permission: "write"
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+  
+  release:
+    name: Publish to Github Packages
+    runs-on: ubuntu-latest
+    needs: [authorize]
+    env:
+      PACKAGE_VERSION: ${{ github.event.inputs.packageVersion }}
+      DIST_TAG: ${{ github.event.inputs.distTag }}
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Use Node.js
+        uses: actions/setup-node@v2
+        with:
+          node-version-file: ".nvmrc"
+          cache: 'yarn'
+
+      - name: Install
+        run: yarn install --frozen-lockfile
+
+      - name: Set registry url
+        uses: actions/setup-node@v2
+        with:
+          registry-url: 'https://registry.npmjs.org'
+      
+      - name: Update version
+        run: yarn version --${{ env.PACKAGE_VERSION }}
+
+      - name: Build npm pkge
+        run: yarn build:prod
+
+      - name: Publish
+        run: npm publish --tag ${{ env.DIST_TAG }}
+        env:
+          NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
\ No newline at end of file
diff --git a/package.json b/package.json
index 803a2983..edff4c89 100644
--- a/package.json
+++ b/package.json
@@ -29,7 +29,9 @@
     "build-npm": "tsc -p ./tsconfig.build.json",
     "build:dev": "export LINK_API_URL='http://localhost:3000'; yarn build",
     "build:dev:watch": "nodemon -e 'ts,tsx,js,json,css,scss,svg' --ignore 'src/**/*-css.ts' --ignore 'src/**/*-svg.ts' --watch src/ --exec 'yarn build:dev'",
-    "build:prod": "yarn prebuild && yarn build && yarn build-npm && cp ./package.json ./README.md ./LICENSE build/npm && cp -a src/vendor-js build/npm/dist && sed -i.bak 's|  \"private\": true,||g' build/npm/package.json && rm -f build/npm/package.json.bak",
+    "build:clean": "rm -f build/npm/package.json.bak",
+    "build:prod": "yarn prebuild && yarn build && yarn build-npm && cp ./package.json ./README.md ./LICENSE build/npm && cp -a src/vendor-js build/npm/dist && sed -i.bak 's|  \"private\": true,||g' build/npm/package.json",
+    "build:prod-clean": "yarn build:prod && yarn build:clean",
     "lint:types": "tsc --noEmit",
     "lint:prettier": "prettier --check \"{src,__tests__}/**/*.(js|ts|tsx)\"",
     "lint:eslint": "eslint ./src --ext .ts,.tsx",

**Comment by Newman1515** [comment]
From 6ce9a64a842b0462c0687420e61a23a9c7b56474 Mon Sep 17 00:00:00 2001
From: Erin Lee <erin.lee@coinbase.com>
Date: Fri, 25 Mar 2022 17:58:40 -0400
Subject: [PATCH 1/4] [chore] github action to publish

---
 .github/workflows/publish.yml | 59 +++++++++++++++++++++++++++++++++++
 package.json                  |  4 ++-
 2 files changed, 62 insertions(+), 1 deletion(-)
 create mode 100644 .github/workflows/publish.yml

diff --git a/.github/workflows/publish.yml b/.github/workflows/publish.yml
new file mode 100644
index 00000000..4d40f1b6
--- /dev/null
+++ b/.github/workflows/publish.yml
@@ -0,0 +1,59 @@
+name: Publish Package to npm
+
+on:
+  workflow_dispatch:
+    inputs:
+      packageVersion:
+        description: "The version to publish (e.g. patch, minor, major)"
+        required: true
+        default: "tag"
+      distTag:
+        description: "The dist-tag to publish (e.g. latest, beta)"
+        required: true
+        default: "latest"
+
+jobs:
+  authorize:
+    name: Authorize
+    runs-on: ubuntu-latest
+    steps:
+      - name: ${{ github.actor }} permission check to do a release
+        uses: "lannonbr/repo-permission-check-action@2.0.2"
+        with:
+          permission: "write"
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+  
+  release:
+    name: Publish to Github Packages
+    runs-on: ubuntu-latest
+    needs: [authorize]
+    env:
+      PACKAGE_VERSION: ${{ github.event.inputs.packageVersion }}
+      DIST_TAG: ${{ github.event.inputs.distTag }}
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Use Node.js
+        uses: actions/setup-node@v2
+        with:
+          node-version-file: ".nvmrc"
+
+      - name: Set registry url
+        uses: actions/setup-node@v2
+        with:
+          registry-url: 'https://registry.npmjs.org'
+          scope: '@coinbase'
+      
+      - name: Update version
+        run: yarn version --${{ env.PACKAGE_VERSION }}
+
+      - name: Build npm pkge
+        run: yarn build:prod
+
+      - name: Publish
+        run: npm publish --tag ${{ env.DIST_TAG }}
+        env:
+          NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
\ No newline at end of file
diff --git a/package.json b/package.json
index 803a2983..edff4c89 100644
--- a/package.json
+++ b/package.json
@@ -29,7 +29,9 @@
     "build-npm": "tsc -p ./tsconfig.build.json",
     "build:dev": "export LINK_API_URL='http://localhost:3000'; yarn build",
     "build:dev:watch": "nodemon -e 'ts,tsx,js,json,css,scss,svg' --ignore 'src/**/*-css.ts' --ignore 'src/**/*-svg.ts' --watch src/ --exec 'yarn build:dev'",
-    "build:prod": "yarn prebuild && yarn build && yarn build-npm && cp ./package.json ./README.md ./LICENSE build/npm && cp -a src/vendor-js build/npm/dist && sed -i.bak 's|  \"private\": true,||g' build/npm/package.json && rm -f build/npm/package.json.bak",
+    "build:clean": "rm -f build/npm/package.json.bak",
+    "build:prod": "yarn prebuild && yarn build && yarn build-npm && cp ./package.json ./README.md ./LICENSE build/npm && cp -a src/vendor-js build/npm/dist && sed -i.bak 's|  \"private\": true,||g' build/npm/package.json",
+    "build:prod-clean": "yarn build:prod && yarn build:clean",
     "lint:types": "tsc --noEmit",
     "lint:prettier": "prettier --check \"{src,__tests__}/**/*.(js|ts|tsx)\"",
     "lint:eslint": "eslint ./src --ext .ts,.tsx",

From 6f1cee03a5f129dee8740f7731ce1c7135fd06b8 Mon Sep 17 00:00:00 2001
From: Erin Lee <erin.lee@coinbase.com>
Date: Fri, 25 Mar 2022 18:01:54 -0400
Subject: [PATCH 2/4] fix default tag

---
 .github/workflows/publish.yml | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/.github/workflows/publish.yml b/.github/workflows/publish.yml
index 4d40f1b6..16cf7756 100644
--- a/.github/workflows/publish.yml
+++ b/.github/workflows/publish.yml
@@ -6,7 +6,7 @@ on:
       packageVersion:
         description: "The version to publish (e.g. patch, minor, major)"
         required: true
-        default: "tag"
+        default: "patch"
       distTag:
         description: "The dist-tag to publish (e.g. latest, beta)"
         required: true

From 39eed053a4735273725d3f9aa1f8e4da04e475ae Mon Sep 17 00:00:00 2001
From: Erin Lee <erin.lee@coinbase.com>
Date: Fri, 25 Mar 2022 18:15:39 -0400
Subject: [PATCH 3/4] use default

---
 .github/workflows/publish.yml | 1 -
 1 file changed, 1 deletion(-)

diff --git a/.github/workflows/publish.yml b/.github/workflows/publish.yml
index 16cf7756..435209df 100644
--- a/.github/workflows/publish.yml
+++ b/.github/workflows/publish.yml
@@ -45,7 +45,6 @@ jobs:
         uses: actions/setup-node@v2
         with:
           registry-url: 'https://registry.npmjs.org'
-          scope: '@coinbase'
       
       - name: Update version
         run: yarn version --${{ env.PACKAGE_VERSION }}

From 3766c78d1c593e8e77645377fac4da8b81dbeb07 Mon Sep 17 00:00:00 2001
From: Erin Lee <erin.lee@coinbase.com>
Date: Fri, 25 Mar 2022 18:19:47 -0400
Subject: [PATCH 4/4] yarn install step

---
 .github/workflows/publish.yml | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/.github/workflows/publish.yml b/.github/workflows/publish.yml
index 435209df..0fc25bd1 100644
--- a/.github/workflows/publish.yml
+++ b/.github/workflows/publish.yml
@@ -40,6 +40,10 @@ jobs:
         uses: actions/setup-node@v2
         with:
           node-version-file: ".nvmrc"
+          cache: 'yarn'
+
+      - name: Install
+        run: yarn install --frozen-lockfile
 
       - name: Set registry url
         uses: actions/setup-node@v2

File: .github/workflows/publish.yml (hunk 3 above)
Comment (macky22am) [reply .github/workflows/publish.yml:42]: Did I miss something

**Comment by bangtoven** [comment]
closing it since it's been inactive for 180+ days. please reopen if needed.

**Pull request closed without merging**